File: android/src/main/java/com/smileidentity/react/expo/SmileIDDocumentVerificationEnhancedView.kt
package com.smileidentity.react.expo

import android.content.Context
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.systemBarsPadding
import androidx.compose.runtime.Composable
import androidx.compose.runtime.mutableStateOf
import androidx.compose.ui.Modifier
import com.smileidentity.SmileID
import com.smileidentity.compose.EnhancedDocumentVerificationScreen
import com.smileidentity.results.EnhancedDocumentVerificationResult
import com.smileidentity.results.SmileIDResult
import com.smileidentity.util.randomJobId
import com.smileidentity.util.randomUserId
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.viewevent.EventDispatcher

/**
 * Document Verification Enhanced View using ExpoView
 */
class SmileIDDocumentVerificationEnhancedView(context: Context, appContext: AppContext) :
    SmileIDExpoComposeView(
        context = context,
        appContext = appContext,
        shouldHostComposeContent = true
    ) {
    private var props = mutableStateOf(DocumentVerificationProps())
    private val onResult by EventDispatcher()
    private val onError by EventDispatcher()

    fun updateConfig(config: EnhancedDocumentVerificationRequest) {
        props.value = config.toDocumentVerificationProps()
    }

    @Composable
    override fun Content() {
        EnhancedDocumentVerificationView(
            props = props.value,
            onResult = { result ->
                onResult(
                    mapOf(
                        "documentFrontFile" to result.documentFrontFile.toString(),
                        "documentBackFile" to result.documentBackFile.toString(),
                        "selfieFile" to result.selfieFile.toString(),
                    )
                )
            },
            onError = { error ->
                onError(
                    mapOf(
                        "error" to error.localizedMessage.toString()
                    )
                )
            }
        )
    }
}

/**
 * Compose view that wraps the SmileID enhanced document verification view
 **/
@Composable
fun EnhancedDocumentVerificationView(
    props: DocumentVerificationProps,
    onResult: (EnhancedDocumentVerificationResult) -> Unit,
    onError: (Throwable) -> Unit
) {
    Column(
        modifier = Modifier
            .fillMaxSize()
            .navigationBarsPadding()
            .systemBarsPadding()
    ) {
        SmileID.EnhancedDocumentVerificationScreen(
            userId = props.userId ?: randomUserId(),
            jobId = props.jobId ?: randomJobId(),
            allowNewEnroll = props.allowNewEnroll,
            countryCode = props.countryCode,
            documentType = props.documentType,
            idAspectRatio = props.idAspectRatio,
            bypassSelfieCaptureWithFile = props.bypassSelfieCaptureWithFile,
            enableAutoCapture = props.enableAutoCapture,
            captureBothSides = props.captureBothSides,
            allowAgentMode = props.allowAgentMode,
            allowGalleryUpload = props.allowGalleryUpload,
            showInstructions = props.showInstructions,
            showAttribution = props.showAttribution,
            useStrictMode = props.useStrictMode,
            extraPartnerParams = props.extraParams,
            consentInformation = props.consentInformation
        ) { result ->
            when (result) {
                is SmileIDResult.Success -> {
                    onResult(result.data)
                }
                is SmileIDResult.Error -> {
                    onError(result.throwable)
                }
            }

        }
    }
}

File: ios/SmileIDDocumentVerificationEnhancedView.swift
import ExpoModulesCore
import SmileID
import SwiftUI

// Enhanced Document Verification View using ExpoView
final class SmileIDDocumentVerificationEnhancedView: ExpoView {
    let onResult = EventDispatcher()
    let onError = EventDispatcher()
    private let delegate: EnhancedDocumentVerificationDelegate
    private let hostingController: UIHostingController<EnhancedDocumentVerificationView>
    private var config: EnhancedDocumentVerificationRecord?

    required init(appContext: AppContext? = nil) {
        delegate = EnhancedDocumentVerificationDelegate()
        hostingController = UIHostingController(
            rootView: EnhancedDocumentVerificationView(
                delegate: delegate,
                config: nil
            )
        )
        super.init(appContext: appContext)

        // Set up delegate callbacks
        delegate.onResult = { [weak self] result in
            self?.onResult(result)
        }

        delegate.onError = { [weak self] error in
            self?.onError(["error": error.localizedDescription])
        }

        // Add the hosting controller's view
        addSubview(hostingController.view)
        hostingController.view.fillSuperview()
    }

    func updateConfig(_ config: EnhancedDocumentVerificationRecord) {
        self.config = config
        hostingController.rootView = EnhancedDocumentVerificationView(
            delegate: delegate,
            config: config
        )
    }
}

// SwiftUI view that wraps the SmileID enhanced document verification screen
struct EnhancedDocumentVerificationView: View {
    let delegate: EnhancedDocumentVerificationDelegate
    let config: EnhancedDocumentVerificationRecord?

    var body: some View {
        if let config = config {
            SmileID.enhancedDocumentVerificationScreen(
                userId: config.userId ?? generateUserId(),
                jobId: config.jobId ?? generateJobId(),
                allowNewEnroll: config.allowNewEnroll,
                countryCode: config.countryCode,
                documentType: config.documentType,
                idAspectRatio: config.idAspectRatio,
                bypassSelfieCaptureWithFile: config.bypassSelfieCaptureWithFile.flatMap(URL.init),
                enableAutoCapture: config.enableAutoCapture,
                captureBothSides: config.captureBothSides,
                allowAgentMode: config.allowAgentMode,
                allowGalleryUpload: config.allowGalleryUpload,
                showInstructions: config.showInstructions,
                skipApiSubmission: config.skipApiSubmission,
                showAttribution: config.showAttribution,
                useStrictMode: config.useStrictMode,
                extraPartnerParams: config.extraPartnerParams,
                consentInformation: ConsentInformation(
                    consented: ConsentedInformation(
                        consentGrantedDate: config.consentInformationRequest?.consentGrantedDate ?? Date().toISO8601WithMilliseconds(),
                        personalDetails: config.consentInformationRequest?.personalDetails ?? false,
                        contactInformation: config.consentInformationRequest?.contactInformation ?? false,
                        documentInformation: config.consentInformationRequest?.documentInformation ?? false
                    )
                ),
                delegate: delegate
            )
        } else {
            Text("Configuration not provided")
                .foregroundColor(.red)
        }
    }
}


// Delegate class for enhanced document verification
class EnhancedDocumentVerificationDelegate: EnhancedDocumentVerificationResultDelegate {
    var onResult: (([String: Any]) -> Void)?
    var onError: ((Error) -> Void)?

    func didSucceed(
        selfie: URL,
        documentFrontImage: URL,
        documentBackImage: URL?,
        didSubmitEnhancedDocVJob: Bool
    ) {
        var params: [String: Any] = [
            "selfie": selfie.absoluteString,
            "documentFrontFile": documentFrontImage.absoluteString,
            "didSubmitEnhancedDocVJob": didSubmitEnhancedDocVJob,
        ]
        if let documentBackImage {
            params["documentBackFile"] = documentBackImage.absoluteString
        }
        onResult?(params)
    }

    func didError(error: Error) {
        onError?(error)
    }
}
